arch/x86/kvm/pkvm/pkvm.c

```diff
@@ -52,3 +54,50 @@ void pkvm_kick_vcpu(struct kvm_vcpu *vcpu)
 
 	pkvm_lapic_send_init(READ_ONCE(vcpu->cpu));
 }
+
+int pkvm_x86_vendor_init(struct kvm_x86_init_ops *ops)
+{
+	int r;
+
+	memset(&kvm_caps, 0, sizeof(kvm_caps));
+
+	kvm_caps.supported_vm_types = BIT(KVM_X86_DEFAULT_VM) |
+				      BIT(KVM_X86_PKVM_PROTECTED_VM);
+	kvm_caps.supported_mce_cap = MCG_CTL_P | MCG_SER_P;
+
+	if (boot_cpu_has(X86_FEATURE_XSAVE)) {
+		kvm_host.xcr0 = xgetbv(XCR_XFEATURE_ENABLED_MASK);
+		kvm_caps.supported_xcr0 = kvm_host.xcr0 & KVM_SUPPORTED_XCR0;
+	}
+
+	if (boot_cpu_has(X86_FEATURE_XSAVES)) {
+		rdmsrq(MSR_IA32_XSS, kvm_host.xss);
+		kvm_caps.supported_xss = kvm_host.xss & KVM_SUPPORTED_XSS;
+	}
+
+	rdmsrq_safe(MSR_EFER, &kvm_host.efer);
+
+	if (boot_cpu_has(X86_FEATURE_ARCH_CAPABILITIES))
+		rdmsrq(MSR_IA32_ARCH_CAPABILITIES, kvm_host.arch_capabilities);
+
+	r = ops->hardware_setup();
+	if (r)
+		return r;
+
+	memcpy(&kvm_x86_ops, ops->runtime_ops, sizeof(kvm_x86_ops));
+
+	if (!kvm_cpu_cap_has(X86_FEATURE_XSAVES))
+		kvm_caps.supported_xss = 0;
+
+	if (!kvm_cpu_cap_has(X86_FEATURE_SHSTK) &&
+	    !kvm_cpu_cap_has(X86_FEATURE_IBT))
+		kvm_caps.supported_xss &= ~XFEATURE_MASK_CET_ALL;
+
+	if ((kvm_caps.supported_xss & XFEATURE_MASK_CET_ALL) != XFEATURE_MASK_CET_ALL) {
+		kvm_cpu_cap_clear(X86_FEATURE_SHSTK);
+		kvm_cpu_cap_clear(X86_FEATURE_IBT);
+		kvm_caps.supported_xss &= ~XFEATURE_MASK_CET_ALL;
+	}
```

Comment:
Just a note: looks like we quite some duplication with kvm_x86_vendor_init()... But it still might be cleaner than adding a bunch of #ifdefs to kvm_x86_vendor_init() to reuse it.

Reply:
Actually I didn't suggest to change it, I meant that maybe for now it actually would be best keep it as is, i.e. keep a bit of duplication here. (But if you prefer to change it to reuse  kvm_x86_vendor_init() like we did in v6.12, I'm probably fine with that too.)